master/internal/core_checkpoint.go

```diff
@@ -0,0 +1,397 @@
+package internal
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"compress/gzip"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/google/uuid"
+	"github.com/labstack/echo/v4"
+	"github.com/pkg/errors"
+
+	"github.com/determined-ai/determined/master/internal/api"
+	"github.com/determined-ai/determined/master/pkg/ptrs"
+	"github.com/determined-ai/determined/master/pkg/schemas/expconf"
+)
+
+type archiveWriter interface {
+	WriteFileHeader(fname string, size int64) error
+	Write(b []byte) (int, error)
+}
+
+type tarArchiveWriter struct {
+	tw *tar.Writer
+}
+
+func (aw *tarArchiveWriter) WriteFileHeader(fname string, size int64) error {
+	hdr := tar.Header{
+		Name: fname,
+		Mode: 0666,
+		Size: size,
+	}
+	if strings.HasSuffix(fname, "/") {
+		// This a directory
+		hdr.Mode = 0777
+	}
+	return aw.tw.WriteHeader(&hdr)
+}
+
+func (aw *tarArchiveWriter) Write(p []byte) (int, error) {
+	return aw.tw.Write(p)
+}
+
+type zipArchiveWriter struct {
+	zw        *zip.Writer
+	zwContent io.Writer
+}
+
+func (aw *zipArchiveWriter) WriteFileHeader(fname string, size int64) error {
+	// Zip by default sets mode 0666 and 077 for files and folders respectively
+	zwc, err := aw.zw.Create(fname)
+	if err != nil {
+		return err
+	}
+	aw.zwContent = zwc
+	return nil
+}
+
+func (aw *zipArchiveWriter) Write(p []byte) (int, error) {
+	var w io.Writer
+	if aw.zwContent == nil {
+		return 0, nil
+	}
```

Comment:
What is this check for?

(Also it violates the io.Writer interface, which requires that you either return `n == len(p)` or you return a non-nil error)

Reply:
When `WriteFileHeader()` hasn't been called or has failed, `aw.zwContent` would be `nil`. Yeah, I need to return a non-nil error here.

Reply:
It might be worth including a comment saying it's a guard just in case s3 calls this method even after the Next method returns false?

Reply:
Although we only use it with S3 currently, It doesn't need to be limited to S3. Hence I would mention S3 as an example. Now that I think about it, `s3SeqWriterAt` can well be just `seqWriterAt`.